### PR TITLE
RSDK-578 Add client interceptors

### PR DESCRIPTION
**Pull request by maximpertsov**
JIRA: https://viam.atlassian.net/browse/RSDK-578

Required for https://github.com/viamrobotics/rdk/pull/1501

### CODE REVIEW

File: rpc/dial_options.go [hunk 1]
```diff
@@ -52,6 +53,10 @@ type dialOptions struct {
 
 	// stats monitoring on the connections.
 	statsHandler stats.Handler
+
+	// interceptors
+	unaryInterceptor  grpc.UnaryClientInterceptor
```
Comment (cheukt) [inline rpc/dial_options.go:58]: can we make these lists so we can add more than one interceptor?

Comment (maximpertsov) [reply rpc/dial_options.go:58]: it seems like we can do that by passing in multiple dial options with interceptors and then chaining those interceptors. this is the pattern we have for server options

**Review by cheukt — APPROVED** [review]
LGTM

**Review by edaniels — APPROVED** [review]
LGTM if you add tests :)

File: rpc/dialer_test.go [hunk 2]
```diff
@@ -357,3 +358,146 @@ func TestWithDialStatsHandler(t *testing.T) {
...
+
+	test.That(
+		t,
+		interceptedMethods,
+		test.ShouldResemble,
+		[]string{
+			"/proto.rpc.webrtc.v1.SignalingService/OptionalWebRTCConfig",
+			"/proto.rpc.v1.AuthService/Authenticate",
+			"/proto.rpc.examples.echo.v1.EchoService/Echo",
+			"/proto.rpc.v1.AuthService/Authenticate",
```
Comment (maximpertsov) [inline rpc/dialer_test.go:431]: we probably should exempt auth and signaling method like we do for server-side interceptors

Comment (maximpertsov) [reply rpc/dialer_test.go:431]: @edaniels any thoughts on this?

**Pull request merged**